src/mozanalysis/experiments.py

```diff
@@ -1,165 +1,268 @@
+        df = df.filter(F.col("experiment_id") == self.experiment_id)
+        return df
+
+    def aggregate_per_day(self, df, core_metric=None):
+        # If user wants to override outlier percentiles, set it False here and
+        # call it explicitly.
+
+        if core_metric is None:
+            columns = list(set(sum(COLUMNS.values(), [])))
+        else:
+            columns = COLUMNS[core_metric]
 
         df = (
-            dataset.select(*cols)
-            .groupBy(self._aggregate_by, self._split_by, self._date_aggregate_by)
-            .agg(*aggs)
+            df.select(
+                "client_id",
+                "experiment_branch",
+                F.date_format(
+                    F.to_date("submission_date_s3", format="yyyyMMdd"), "yyyy-MM-dd"
+                ).alias("submission_date"),
+                *columns
+            )
+            .filter(F.col("experiment_branch").isNotNull())
+            .groupBy(*BASE_COLS)
+            # This aggregation simply SUMs the metrics per client per branch
+            # per day. Note: This may not work for future metrics that aren't
+            # simple scalars.
+            .agg(
+                *[
+                    F.expr("SUM(COALESCE({}, 0))".format(col)).alias(col)
+                    for col in columns
+                ]
+            )
         )
+
+        return df
+
+    def get_data(self, trim_outliers=True, core_metric=None):
+        df = self.get_experiment_data()
+        df = self.aggregate_per_day(df, core_metric)
+
+        # TODO: Cache this dataframe.
+
+        self._df = df
         return df
 
-    def analyze(self, dataset):
-        """
-        Returns a panda dataframe for the provided metrics.
-
-        This also splits the dataset into multiple groups using the column
-        defined in the `split_by` call, which is "experiment_branch" by
-        default.
-
-        Arguments:
-
-        - dataset (a pyspark Dataframe) : Expects a dataframe with the columns
-            defined in the `MetricDefinition`s `columns` field.
-
-        """
-        dataset_was_cached = False
-        # `DataFrame.is_cached` is undocumented, so try to avoid breaking
-        # if a Spark upgrade drops it.
-        try:
-            dataset_was_cached = dataset.is_cached
-        except AttributeError:
-            warnings.warn(
-                "Could not determine whether the provided DataFrame was already "
-                "cached. `analyze` will unpersist the DataFrame. "
-                "(This does not affect the analysis.)"
+    def trim_outliers(self, outlier_percentile=0.9999, relative_error=0.0001):
+        # TODO
+        # Updates `self._df` and returns it.
+        return self._df
+
+    def engagement(self):
+        # Compute only the engagement metrics.
+        # Return pandas dataframe with data and CIs.
+        if not self._df:
+            self._df = self.get_data(metric=CoreMetrics.ENGAGEMENT)
```

Comment:
What if `self._df` has the wrong metrics?

(Suggestion: either store the flag that `get_data` was passed and make a decision based on that, or else make calculating retention the responsibility of a second class)

Reply:
I like the idea of storing some extra metadata about what's saved in the `_df`.